rp2040-hal/src/gpio.rs

```diff
@@ -1,132 +1,337 @@
+                        self,
+                    ) -> $PXi<Output> {
+                        unsafe {
+                            setup_pad_io(&*pac::$PADSX::ptr(), $i);
+                        }
+                        unsafe {
+                            set_gpio_function(&*pac::$GPIOX::ptr(), $i, FUNCTION_SIO);
+                        }
+                        unsafe {
+                            (*pac::SIO::ptr()).gpio_oe_set.write(|x| { x.bits(1 << $i) });
+                        }
+                        $PXi { _mode: PhantomData }
+                    }
+
+                    #[doc = "Configure this pin as an input"]
+                    pub fn into_input(
+                        self,
+                    ) -> $PXi<Input> {
+                        unsafe {
+                            setup_pad_io(&*pac::$PADSX::ptr(), $i);
+                        }
+                        unsafe {
+                            set_gpio_function(&*pac::$GPIOX::ptr(), $i, FUNCTION_SIO);
+                        }
+                        unsafe {
+                            (*pac::SIO::ptr()).gpio_oe_clr.write(|x| { x.bits(1 << $i) });
+                        }
+                        $PXi { _mode: PhantomData }
                     }
-                    sio.gpio_oe.modify(|_, x| unsafe { x.bits(1 << $i) });
-                    $PXi { _mode: PhantomData }
                 }
-            }
 
-            impl OutputPin for $PXi<Output> {
-                type Error = Infallible;
+                impl OutputPin for $PXi<Output> {
+                    type Error = Infallible;
+
+                    fn set_low(&mut self) -> Result<(), Self::Error> {
+                        unsafe {
+                            (*pac::SIO::ptr()).gpio_out_clr.write(|x| x.bits(1 << $i));
+                        }
+                        Ok(())
+                    }
 
-                fn set_low(&mut self) -> Result<(), Self::Error> {
-                    unsafe {
-                        (*pac::SIO::ptr()).gpio_out_clr.write(|x| x.bits(1 << $i));
+                    fn set_high(&mut self) -> Result<(), Self::Error> {
+                        unsafe {
+                            (*pac::SIO::ptr()).gpio_out_set.write(|x| x.bits(1 << $i));
+                        }
+                        Ok(())
                     }
-                    Ok(())
                 }
 
-                fn set_high(&mut self) -> Result<(), Self::Error> {
-                    unsafe {
-                        (*pac::SIO::ptr()).gpio_out_set.write(|x| x.bits(1 << $i));
+                impl StatefulOutputPin for $PXi<Output> {
+                    fn is_set_low(&self) -> Result<bool, Self::Error> {
+                        Ok(!self.is_set_high()?)
+                    }
+
+                    fn is_set_high(&self) -> Result<bool, Self::Error> {
+                        unsafe {
+                            Ok((*pac::SIO::ptr()).gpio_out_set.read().bits() & (1 << $i) != 0)
```

Comment:
Minor: The read value of the atomic registers is not really described in the datasheet so I would just use the "normal" register not the `set` one

```suggestion
                            Ok((*pac::SIO::ptr()).gpio_out.read().bits() & (1 << $i) != 0)
```

Reply:
:woman_facepalming: this is why my input example stopped working. I thought I had just broken my board or something.